contracts/0.8.9/WithdrawalQueueBase.sol

```diff
@@ -0,0 +1,379 @@
+        discountHistory[lastDiscountIndex] = Discount(lastRequestId, NO_DISCOUNT);
+    }
+
+    /// @notice return the number of unfinalized requests in the queue
+    function unfinalizedRequestNumber() external view returns (uint256) {
+        return lastRequestId - lastFinalizedRequestId;
+    }
+
+    /// @notice Returns the amount of stETH yet to be finalized
+    function unfinalizedStETH() external view returns (uint256) {
+        return queue[lastRequestId].cumulativeStETH - queue[lastFinalizedRequestId].cumulativeStETH;
+    }
+
+    /**
+     * @notice Returns all withdrawal requests placed for the `_recipient` address
+     *
+     * WARNING: This operation will copy the entire storage to memory, which can be quite expensive. This is designed
+     * to mostly be used by view accessors that are queried without any gas fees. Developers should keep in mind that
+     * this function has an unbounded cost, and using it as part of a state-changing function may render the function
+     * uncallable if the set grows to a point where copying to memory consumes too much gas to fit in a block.
+     */
+    function getWithdrawalRequests(address _recipient) external view returns (uint256[] memory requestsIds) {
+        return requestsByRecipient[_recipient].values();
+    }
+
+    /**
+     * @notice Returns status of the withdrawal request
+     */
+    function getWithdrawalRequestStatus(uint256 _requestId)
+        external
+        view
+        returns (
+            uint256 amountOfStETH,
+            uint256 amountOfShares,
+            address recipient,
+            uint256 timestamp,
+            bool isFinalized,
+            bool isClaimed
+        )
+    {
+        if (_requestId == 0) revert ZeroRequestId();
+        if (_requestId > lastRequestId) revert InvalidRequestId(_requestId);
+
+        WithdrawalRequest memory request = queue[_requestId];
+        WithdrawalRequest memory previousRequest = queue[_requestId - 1];
+
+        recipient = request.recipient;
+        timestamp = request.timestamp;
+
+        amountOfShares = request.cumulativeShares - previousRequest.cumulativeShares;
+        amountOfStETH = request.cumulativeStETH - previousRequest.cumulativeStETH;
+
+        isFinalized = _requestId <= lastFinalizedRequestId;
+        isClaimed = request.claimed;
+    }
+
+    /**
+     * @notice Returns the amount of ETH to be send along to finalize this batch and the amount of shares to burn after
+     * @param _lastRequestIdToFinalize the index in the request queue that should be used as the end of the batch. Should be >= 1
+     * @param _shareRate share rate that will be used to calculate the batch value
+     *
+     * @return eth amount of ETH required to finalize the batch
+     * @return shares amount of shares that should be burned on finalization
+     */
+    function finalizationBatch(uint256 _lastRequestIdToFinalize, uint256 _shareRate)
```

Comment:
Looks like we need to inline all math because of mul and div on `E27_PRECISION_BASE `